File: December 2024 Leetcode Solution/December-03.java
class Solution {
    public String addSpaces(String s, int[] spaces) {
        int n = s.length();
        int m = spaces.length;
        StringBuilder result = new StringBuilder(n + m);

        int spaceIndex = 0, strIndex = 0;
        for (int i = 0; i < n; i++) {
            if (spaceIndex < m && i == spaces[spaceIndex]) {
                result.append(' ');
                spaceIndex++;
            }
            result.append(s.charAt(i));
        }

        return result.toString();
    }
}

2)
class Solution {
    public String addSpaces(String s, int[] spaces) {
        StringBuilder result = new StringBuilder(s.length() + spaces.length);
        int spaceIndex = 0;

        for (int i = 0; i < s.length(); i++) {
            if (spaceIndex < spaces.length && i == spaces[spaceIndex]) {
                result.append(' ');
                spaceIndex++;
            }
            result.append(s.charAt(i));
        }

        return result.toString();
    }
}

File: December 2024 Leetcode Solution/December-03.py
class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        spaces.append(len(s))
        return ''.join(s[spaces[i - 1] if i > 0 else 0:spaces[i]] + (' ' if i < len(spaces) - 1 else '') for i in range(len(spaces)))

2)
class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        result = []
        prev = 0
        
        for space in spaces:
            result.append(s[prev:space])
            result.append(" ")
            prev = space
        
        result.append(s[prev:])
        return ''.join(result)

3)
class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        n = len(s)
        result = []
        prev = 0
        spaces_idx = 0
        spaces_len = len(spaces)
        for i in range(n):
            if spaces_idx < spaces_len and i == spaces[spaces_idx]:
                result.append(" ")
                spaces_idx += 1
            result.append(s[i])
        
        return ''.join(result)

4)
class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        result = []
        spaceIndex = 0
        for i in range(len(s)):
            if spaceIndex < len(spaces) and i == spaces[spaceIndex]:
                result.append(' ')
                spaceIndex += 1
            result.append(s[i])
        return ''.join(result)

5)
class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        result = []
        space_index = 0

        for i, char in enumerate(s):
            if space_index < len(spaces) and i == spaces[space_index]:
                result.append(' ')
                space_index += 1
            result.append(char)

        return ''.join(result)
